### ISSUE TITLE
Thinkpad T490: TPM interrupt not working, polling instead; unsealing key fails

### ISSUE THREAD

**Issue by sidhussmann** [osresearch/safeboot#93]
This issue picks up on a discussion in https://github.com/osresearch/safeboot/issues/91#issuecomment-738460669

On Thinkpad T490 I'm unable to unseal the LUKS key. It seems as if there is something wrong with the TPM firmware:
```
[    1.085251] tpm_tis NTC0702:00: 2.0 TPM (device-id 0xFC, rev-id 1)
[    1.085433] tpm tpm0: tpm_try_transmit: send(): error -5
[    1.085438] tpm tpm0: [Firmware Bug]: TPM interrupt not working, polling instead
```

The firmware is up-to-date and has following device tree:
```
$ sudo fwupdmgr get-devices --show-all-devices
20N2CTO1WW
│
├─System Firmware:
│     Device ID:           6150dd1f7291b0709289ab8a53cc85a17e117ef2
│     Current version:     0.1.70
│     Minimum Version:     0.0.1
│     Vendor:              LENOVO (DMI:LENOVO)
│     GUID:                603baf73-b997-45b5-86b4-2f981a008e18
│     Device Flags:        • Internal device
│                          • Updatable
│                          • Requires AC power
│                          • Needs a reboot after installation
│                          • Cryptographic hash verification is available
│                          • Device is usable for the duration of the update
│   
├─TPM:
│ │   Device ID:           c6a80ac3a22083423992a3cb15018989f37834d6
│ │   Current version:     7.2.1.0
│ │   Vendor:              Nuvoton Technology (TPM:NTC)
│ │   GUIDs:               ff71992e-52f7-5eea-94ef-883e56e034c6 ← system-tpm
│ │                        fac1c8f3-73c8-5cd6-8330-07a3690b5140 ← TPM\VEN_NTC&DEV_0000
│ │                        8d94eb9a-78aa-513d-af50-319e3b9f5210 ← TPM\VEN_NTC&MOD_NPCT75x$rls
│ │                        e9ccc1dc-960a-5e09-afe9-e59a904b776d ← TPM\VEN_NTC&DEV_0000&VER_2.0
│ │                        f1c4013d-3f4c-5623-a40b-81ad96e23860 ← TPM\VEN_NTC&MOD_NPCT75x$rls&VER_2.0
│ │   Device Flags:        • Internal device
│ │ 
│ └─Event Log:
│       Device ID:         58bd405f31c48e6eca290b425f530a94c91e955c
│       Vendor:            Nuvoton Technology (TPM:NTC)
│       GUID:              a25657fe-b5dc-5be0-8b78-8b9dfec678ff ← system-tpm-eventlog
│       Device Flags:      • Internal device
│     
├─UEFI Device Firmware:
│     Device ID:           b1c32d855d85a2124e865f0a03f8c2384dcca167
│     Current version:     192.47.1524
│     Minimum Version:     192.47.1524
│     Vendor:              DMI:LENOVO
│     GUID:                bf2d0f8b-f9a9-400c-8914-36c225d16eb4
│     Device Flags:        • Internal device
│                          • Updatable
│                          • Requires AC power
│                          • Needs a reboot after installation
│                          • Device is usable for the duration of the update
│   
├─UEFI Device Firmware:
│     Device ID:           489f23b2ba9c1adf3e9f9f10598c98ba5c6bba39
│     Current version:     0.1.19
│     Minimum Version:     0.1.19
│     Vendor:              DMI:LENOVO
│     GUID:                38ea6335-29ca-417b-8cd4-6b4e5e866f92
│     Device Flags:        • Internal device
│                          • Updatable
│                          • Requires AC power
│                          • Needs a reboot after installation
│                          • Device is usable for the duration of the update
│   
├─UEFI Device Firmware:
│     Device ID:           80992dc07896ebf6cc1be3bb82752367deee6182
│     Current version:     1.1.8343
│     Minimum Version:     0.0.1
│     Vendor:              DMI:LENOVO
│     GUID:                24571d7c-42a4-4739-afc5-0f529b9581b9
│     Device Flags:        • Internal device
│                          • Updatable
│                          • Requires AC power
│                          • Needs a reboot after installation
│                          • Device is usable for the duration of the update
│   
├─UEFI Device Firmware:
│     Device ID:           f60653d31d92c3e457bf7e5606fe7dda211b72d4
│     Current version:     7.2.256
│     Minimum Version:     7.2.256
│     Vendor:              DMI:LENOVO
│     GUID:                bb667386-6ec6-4d32-bb8c-5183a6cc2c95
│     Device Flags:        • Internal device
│                          • Updatable
│                          • Requires AC power
│                          • Needs a reboot after installation
│                          • Device is usable for the duration of the update
│   
├─UHD Graphics 620 (Whiskey Lake):
│     Device ID:           bbbf1ce3d1cf15550c3760b354592040292415bb
│     Current version:     02
│     Vendor:              Intel Corporation (PCI:0x8086)
│     GUIDs:               6986c21a-0f3f-5538-bcef-fe4e8cda7ab7 ← PCI\VEN_8086&DEV_3EA0&REV_02
│                          52aa8ac7-c837-561f-af32-b7979cd2f6ed ← PCI\VEN_8086&DEV_3EA0
│     Device Flags:        • Internal device
│                          • Cryptographic hash verification is available
│   
└─WDC PC SN720 SDAQNTW-256G-1001:
      Device ID:           f2759da7fe8e0388c5f3601cb072f837b1070b03
      Summary:             NVM Express Solid State Drive
      Current version:     10190101
      Vendor:              Sandisk Corp (NVME:0x15B7)
      Serial Number:       19285E429922
      GUIDs:               137520ce-3603-53e6-9165-56694ed744e7 ← NVME\VEN_15B7&DEV_5002&REV_00
                           ff2112dc-038c-596d-90ca-d43c5077c6ec ← NVME\VEN_15B7&DEV_5002
                           237776ee-0bcd-5fe9-8dc8-6984a2d36ba0 ← WDC PC SN720 SDAQNTW-256G-1001
      Device Flags:        • Internal device
                           • Updatable
                           • Requires AC power
                           • Needs a reboot after installation
                           • Device is usable for the duration of the update
```

On Ubuntu 20.04.1 LTS (GNU/Linux 5.4.0-56-generic x86_64) and safeboot release-0.7, after going through the steps described in https://github.com/osresearch/safeboot/issues/91#issuecomment-734347930 the final steps in TL:TR produce following output:

```
root@sid-ThinkPad-T490:/home/sid# safeboot linux-sign                        
/dev/mapper/vgubuntu-root on / type ext4 (ro,relatime)
/dev/mapper/vgubuntu-root: already read-only
/dev/mapper/vgubuntu-root: Computing hashes: this will take a while...
# cryptsetup 2.2.2 processing "veritysetup format --debug /dev/mapper/vgubuntu-root /dev/mapper/vgubuntu-hashes"
# Running command format.
# Allocating context for crypt device /dev/mapper/vgubuntu-hashes.
# Trying to open and read device /dev/mapper/vgubuntu-hashes with direct-io.
# Initialising device-mapper backend library.
# Formatting device /dev/mapper/vgubuntu-hashes as type VERITY.
# Crypto backend (OpenSSL 1.1.1f  31 Mar 2020) initialized in cryptsetup library version 2.2.2.
# Detected kernel Linux 5.4.0-56-generic x86_64.
# Setting ciphertext data device to /dev/mapper/vgubuntu-root.
# Trying to open and read device /dev/mapper/vgubuntu-root with direct-io.
# Hash creation sha256, data device /dev/mapper/vgubuntu-root, data blocks 3145728, hash_device /dev/mapper/vgubuntu-hashes, offset 1.
# Using 4 hash levels.
# Data device size required: 12884901888 bytes.
# Hash device size required: 101466112 bytes.
# Updating VERITY header of size 512 on device /dev/mapper/vgubuntu-hashes, offset 0.
VERITY header information for /dev/mapper/vgubuntu-hashes
UUID:                   0633f0a7-aaa8-404a-b702-c45d4cac7e23
Hash type:              1
Data blocks:            3145728
Data block size:        4096
Hash block size:        4096
Hash algorithm:         sha256
Salt:                   6939232c2a3c9a24397f11b15482ce17c87a98b0361bf4fc7f1e4afd3e6b5528
Root hash:              0c0b81522d3e125e4871e2668526a9b927ce785ba5e601e8cff056435dfc25e7
# Releasing crypt device /dev/mapper/vgubuntu-hashes context.
# Releasing device-mapper backend.
# Closing read write fd for /dev/mapper/vgubuntu-hashes.
Command successful.
Boot0003* linux
Kernel commandline: 'ro quiet splash vt.handoff=7 intel_iommu=on efi=disable_early_pci_dma lockdown=confidentiality                     root=/dev/mapper/vroot                  fsck.mode=skip                  verity.hashdev=/dev/mapper/vgubuntu-hashes                         verity.rootdev=/dev/mapper/vgubuntu-root                        verity.hash=0c0b81522d3e125e4871e2668526a9b927ce785ba5e601e8cff056435dfc25e7             safeboot.mode=linux'
/tmp/tmp.x5oMaM4ClP/linux.efi: Creating merged Linux/initrd image
/boot/efi/EFI/linux/linux.efi: Signing (ignore warnings about gaps)
warning: data remaining[108940288 vs 108949900]: gaps between PE/COFF sections?
warning: data remaining[108940288 vs 108949904]: gaps between PE/COFF sections?
Enter engine key pass phrase:
Enter PKCS#11 key PIN for SIGN key:
Signing Unsigned original image
461ca688dbc0ac808801ecb73d63317ebe03816ddeb923a09fa70ed624536787  /boot/efi/EFI/linux/linux.efi
root@sid-ThinkPad-T490:/home/sid# reboot
root@sid-ThinkPad-T490:/home/sid# sudo safeboot luks-seal
Starting PCSCD for yubikey support
!!! /etc/initramfs-tools/hooks/tpm-unseal is not installed; this will probably not work
00000000 [139717432518592] utils.c:81:GetDaemonPid() Can't open /run/pcscd/pcscd.pid: No such file or directory
New unsealing PIN: 03124189 [139717432518592] ccid_usb.c:1264:ControlUSB() control failed (1/2): -7 LIBUSB_ERROR_TIMEOUT
00012339 [139717432518592] ifdhandler.c:150:CreateChannelByNameOrChannel() failed
00000036 [139717432518592] readerfactory.c:1105:RFInitializeReader() Open Port 0x200001 Failed (usb:1050/0407:libudev:0:/dev/bus/usb/001/003)
00000011 [139717432518592] readerfactory.c:376:RFAddReader() Yubico YubiKey OTP+FIDO+CCID init failed.
00002547 [139717432518592] ifdhandler.c:150:CreateChannelByNameOrChannel() failed
00000009 [139717432518592] readerfactory.c:1105:RFInitializeReader() Open Port 0x200001 Failed (usb:1050/0407:libudev:1:/dev/bus/usb/001/003)
00000005 [139717432518592] readerfactory.c:376:RFAddReader() Yubico YubiKey OTP+FIDO+CCID init failed.

Unsealing PIN again: 
Using placeholder TPM counter version
Sealing secret with TPM, storing sealed secret in 0x1500010
persistent-handle: 0x81110000
action: evicted
persistent-handle: 0x81110000
action: persisted
/dev/nvme0n1p3: Current recovery password: 
Removing old LUKS TPM key (if it exists)
Adding new LUKS TPM key
/dev/nvme0n1p3: sealed with PCR 0,2,4,5,7,14
-------- Need to sign PCR and counter values --------
/boot/efi/EFI/linux/linux.efi: TPM version 0000000000000014
sha256:
  0 : 0x4C800880F21A302C6F4ECBE005031C8407B68332A36D6AE6146E9A525CC0E102
  2 : 0x3D458CFE55CC03EA1F443F1562BEEC8DF51C75E14A9FCF9A7234A13F198E7969
  4 : 0x34ABAC820388B9AA3B0741BB6971C7D0D80440668B84BAC662602F0BFA52AB94
  5 : 0xB67B849FDDE59ED374B06AA60A8C0DEB43F4E76A0C17B7D1ED68E990603D2613
  7 : 0x65CAF8DD1E0EA7A6347B635D2B379C93B9A1351EDC2AFC3ECDA700E534EB3068
warning: data remaining[108941792 vs 108951408]: gaps between PE/COFF sections?
/boot/efi/EFI/linux/linux.efi: PE hash 3f57be1b26a172a9a64a24049a8252964eaaaaa72fe5420f4917fd7ed12cf4a8
PCR4 cf3bfb9a833756da69057aea25aed4318e22fdfe0f18ea947d95da377d0ba219
mode=linux PCR14=4cc49932dc91c7021e5cfee8231a5f2da3ac1de6df7a7aeff333cc8dfd230f28
Using TPM counter 0000000000000014
engine "pkcs11" set.
Enter PKCS#11 token PIN for test:
Enter PKCS#11 key PIN for SIGN key:
/sys/firmware/efi/efivars/SafebootPCR-8620893e-c793-457e-8a02-41fc83eef3ce: writing new value
/tmp/tmp.YzZ0717Rdt: Unmounting
sid@sid-ThinkPad-T490:~$ sudo reboot
```

At this stage the the disk should unlock automatically. However, this does not happen. The LUKS key is asked during boot with the error message: `TPM TOTP FAILED linux`

**Comment by sidhussmann** [comment]
Reply to https://github.com/osresearch/safeboot/issues/91#issuecomment-738862134 :

@osresearch I did update my firmware before starting the safeboot experiments in November and again yesterday. Could you run `fwupdmgr get-devices --show-all-devices` and `uname -r` on your T490 so I can cross-check the firmware versions? 

Looking at [1], the error message `tpm tpm0: [Firmware Bug]: TPM interrupt not working, polling instead` is known...

[1] https://bugzilla.kernel.org/show_bug.cgi?id=204121

**Comment by sidhussmann** [comment]
Reply to https://github.com/osresearch/safeboot/issues/91#issuecomment-738739240 : 

@hpc4fun, thank you for further explaining the verification of the UUID's. I re-ran the whole setup again using safeboot v0.7, and the three targets `ubuntu`, `Linux-Firmware-Updater`, and `recovery` all show the same UUID. So I assume the `safeboot recovery-sign` does the right thing...

About you writing a detailed description, that would be helpful indeed. However, since this is quite some work on your side, I'm not sure github issues are the right place for that...

I also strongly support your idea to add assumption checking at a few places in the script!

**Comment by osresearch** [comment]
It looks like my Thinkpad T490 model is a 20N3000KMH and has a STM7308 TPM, while yours is a 20N2CTO1WW and has a Nuvoton NTC0702. Mine also has the `Firmware Bug` polling issue; I hadn't noticed it in the dmesg before:
```
efi:  TPMFinalLog=0x7fca2000  SMBIOS=0x7de94000  SMBIOS 3.0=0x7de87000  ACPI=0x7fd0e000  ACPI 2.0=0x7fd0e014  MEMATTR=0x789b4018  ESRT=0x7c282000  TPMEventLog=0x73989018   
ACPI: SSDT 0x000000007DCA6000 000612 (v02 LENOVO Tpm2Tabl 00001000 INTL 20160527)
ACPI: TPM2 0x000000007DCA5000 000034 (v04 LENOVO TP-N2I   00001390 PTEC 00000002)
tpm_tis STM7308:00: 2.0 TPM (device-id 0x0, rev-id 78)
tpm tpm0: tpm_try_transmit: send(): error -5
tpm tpm0: [Firmware Bug]: TPM interrupt not working, polling instead
```

I also need to apologize -- I didn't read through the issue fully and jumped to conclusion on the reason for the failure.  The 0.7 release didn't have the `tpm2-totp` setup instructions, so that is why the TOTP code fails.  Is the device asking for the "recovery code" or the "unseal code"?

At the key entry stage, you can hit Escape to switch to the text console and it should display the TPM PCRs that have been read for that boot.  We can also attempt to use the TPM event log to diagnose why it is not producing the correct values; can you attach the output the parsed event log for both the boot when you signed the hashes and the attempted boot (after providing the disk encryption key)?
```
sudo cat /sys/kernel/security/tpm0/binary_bios_measurements > /tmp/events.bin
tpm2 eventlog /tmp/events.bin
```

My guess is that we are not precomputing the PCR4 correctly for the `mode=linux` boot.

Also, I noticed this warning:
```
/etc/initramfs-tools/hooks/tpm-unseal is not installed; this will probably not work
```
This was an outdated check and was fixed in https://github.com/osresearch/safeboot/commit/e033470c45ad1d0ece7ef2e12164af950fcd5ba4

**Comment by osresearch** [comment]
[20N3000KMH](https://psref.lenovo.com/Detail/ThinkPad/ThinkPad_T490?M=20N3000KMH): Region WE/Netherlands. I can't find the 20N2CTO1WW on the site, although I see `O1WW` models of other thinkpads in other regions. Do they customize the TPM manufacturer based on the destination country?

**Comment by sidhussmann** [comment]
No worries! I can relate :)

> Is the device asking for the "recovery code" or the "unseal code"?

The device is asking for the "recovery key".

> Also, I noticed this warning:
> `/etc/initramfs-tools/hooks/tpm-unseal is not installed; this will probably not work`
> This was an outdated check and was fixed in e033470

This output was a result of the pre-built 0.7 release, recommended by @hpc4fun in https://github.com/osresearch/safeboot/issues/91#issuecomment-734347930. However I also tried with master + my minor patches [1], which includes e033470, and has different outputs. 

> We can also attempt to use the TPM event log to diagnose why it is not producing the correct values; can you attach the output the parsed event log for both the boot when you signed the hashes and the attempted boot (after providing the disk encryption key)?

To limit complexity and confusion I would setup my device using [1] without dm-verity, before investigating. Do you agree? Or would you rather recommend investigating my current setup (release 0.7 with dm-verity)?

[1] https://github.com/sidhussmann/safeboot/tree/first-experiments

**Comment by sidhussmann** [comment]
> [20N3000KMH](https://psref.lenovo.com/Detail/ThinkPad/ThinkPad_T490?M=20N3000KMH): Region WE/Netherlands. I can't find the 20N2CTO1WW on the site, although I see `O1WW` models of other thinkpads in other regions. Do they customize the TPM manufacturer based on the destination country?

Huh, mine was purchased in Switzerland. I sure hope they don't chose the TPM manufacturer based on the destination country because of political reasons ;)

**Comment by sidhussmann** [comment]
@osresearch on the setup with pre-compiled release 0.7 and with dm-verity, I have following results:

> At the key entry stage, you can hit Escape to switch to the text console and it should display the TPM PCRs that have been read for that boot. 

![signal-2020-12-08-130746](https://user-images.githubusercontent.com/37871887/101482201-a1d51880-3956-11eb-9cd5-e4dfdc7a7300.jpeg)

I notice a signature verification error... Might that be related to the Yubikey error messages I get when doing `luks-seal`?:

```
00000000 [140415219189696] utils.c:81:GetDaemonPid() Can't open /run/pcscd/pcscd.pid: No such file or directory
New unsealing PIN: 03125757 [140415219189696] ccid_usb.c:1264:ControlUSB() control failed (1/2): -7 LIBUSB_ERROR_TIMEOUT
00011281 [140415219189696] ifdhandler.c:150:CreateChannelByNameOrChannel() failed
00000037 [140415219189696] readerfactory.c:1105:RFInitializeReader() Open Port 0x200001 Failed (usb:1050/0407:libudev:0:/dev/bus/usb/001/003)
00000012 [140415219189696] readerfactory.c:376:RFAddReader() Yubico YubiKey OTP+FIDO+CCID init failed.
00002299 [140415219189696] ifdhandler.c:150:CreateChannelByNameOrChannel() failed
00000009 [140415219189696] readerfactory.c:1105:RFInitializeReader() Open Port 0x200001 Failed (usb:1050/0407:libudev:1:/dev/bus/usb/001/003)
00000004 [140415219189696] readerfactory.c:376:RFAddReader() Yubico YubiKey OTP+FIDO+CCID init failed.
```

> We can also attempt to use the TPM event log to diagnose why it is not producing the correct values; can you attach the output the parsed event log for both the boot when you signed the hashes and the attempted boot (after providing the disk encryption key)?


After attempted boot:
[tpm-eventlog-after-attempted-boot.log](https://github.com/osresearch/safeboot/files/5659232/tpm-eventlog-after-attempted-boot.log)


At this stage I can no longer do a `linux-sign`:
```
sid@sid-ThinkPad-T490:~$ sudo safeboot linux-sign
/dev/mapper/vgubuntu-root: Computing hashes: this will take a while...
Input/output error while creating hash area.
# cryptsetup 2.2.2 processing "veritysetup format --debug /dev/mapper/vgubuntu-root /dev/mapper/vgubuntu-hashes"
# Running command format.
# Allocating context for crypt device /dev/mapper/vgubuntu-hashes.
# Trying to open and read device /dev/mapper/vgubuntu-hashes with direct-io.
# Initialising device-mapper backend library.
# Formatting device /dev/mapper/vgubuntu-hashes as type VERITY.
# Crypto backend (OpenSSL 1.1.1f  31 Mar 2020) initialized in cryptsetup library version 2.2.2.
# Detected kernel Linux 5.4.0-56-generic x86_64.
# Setting ciphertext data device to /dev/mapper/vgubuntu-root.
# Trying to open and read device /dev/mapper/vgubuntu-root with direct-io.
# Hash creation sha256, data device /dev/mapper/vgubuntu-root, data blocks 3145728, hash_device /dev/mapper/vgubuntu-hashes, offset 1.
# Using 4 hash levels.
# Data device size required: 12884901888 bytes.
# Hash device size required: 101466112 bytes.
# Cannot write digest to hash device.
# Releasing crypt device /dev/mapper/vgubuntu-hashes context.
# Releasing device-mapper backend.
Command failed with code -1 (wrong or missing parameters).
/dev/mapper/vgubuntu-root: unable to compute hashes
```

**Comment by osresearch** [comment]
Eyeballing the PCR hashes look good.  4 was correctly predicted (it involves the PE hash of the unified kernel image), so that's a good sign.

> I notice a signature verification error... Might that be related to the Yubikey error messages I get when doing luks-seal?:

I'm not sure why those errors are output, especially since it seems that openssl is talking to the yubikey correctly:

```
engine "pkcs11" set.
Enter PKCS#11 token PIN for test:
Enter PKCS#11 key PIN for SIGN key:
/sys/firmware/efi/efivars/SafebootPCR-8620893e-c793-457e-8a02-41fc83eef3ce: writing new value
```

Did you start with a softare key and then migrate to a yubikey key? That could explain the signature verification error if there is an old public key in `/etc/safeboot/cert.pem` that is different than the one in your yubikey. `sudo safeboot yubikey-pubkey` should read it from the key and update the files in `/etc/safeboot`.


> At this stage I can no longer do a linux-sign:

Which is by design - the hash storage device is read only since any modifications would cause an dmverity merkle tree failure. We should detect that this is a SIP boot and reuse the existing dmverity hashes by setting `$HASHFILE` instead of requiring the user to pass it in.

**Comment by sidhussmann** [comment]
> Did you start with a softare key and then migrate to a yubikey key? That could explain the signature verification error if there is an old public key in `/etc/safeboot/cert.pem` that is different than the one in your yubikey. `sudo safeboot yubikey-pubkey` should read it from the key and update the files in `/etc/safeboot`.

No. Each iteration is clean-slate. Fresh Ubuntu and fresh safeboot install. And always with Yubikey. The public keys that the Yubikey returns with `yubikey-pubkey` are the same as the ones in `/etc/safeboot/`:

```
sid@sid-ThinkPad-T490:/etc/safeboot$ cd /tmp
sid@sid-ThinkPad-T490:/tmp$ sudo safeboot yubikey-pubkey cert.pem
Extracting signing cert from yubikey
cert.crt: Creating from cert.pem
cert.pub: Creating from cert.pem
sid@sid-ThinkPad-T490:/tmp$ sha256sum cert.crt  cert.pem  cert.pub
ba9cb550e69146c111821e8d0442b6c33c753f120fc0a869e3ea2cf41a38fcc7  cert.crt
0a76657d6752bb745505f2fe78a7a8e12d076d71d65570a3e63160ef8fbb310c  cert.pem
5665847ec8504c491a0baeb611dd96cbc37b7706fd89cd67eb9e83854a1b5e61  cert.pub
sid@sid-ThinkPad-T490:/tmp$ cd -
/etc/safeboot
sid@sid-ThinkPad-T490:/etc/safeboot$ sha256sum cert.crt  cert.pem  cert.pub
ba9cb550e69146c111821e8d0442b6c33c753f120fc0a869e3ea2cf41a38fcc7  cert.crt
0a76657d6752bb745505f2fe78a7a8e12d076d71d65570a3e63160ef8fbb310c  cert.pem
5665847ec8504c491a0baeb611dd96cbc37b7706fd89cd67eb9e83854a1b5e61  cert.pub
```

**Comment by sidhussmann** [comment]
I believe I have found the culprit:

``` 
Unable to verify PCR signature
```

Not sure how I could miss this. It seems that the error is not about verifying a signature the yubikey created. It's about verifying the signature the TPM created. There doesn't seem to be any Nuvoton certificates in [1], which could be an explanation why the signature verification fails for my machine, which has a Nuvoton NTC0702 TPM...

[1] https://github.com/osresearch/safeboot/tree/master/certs

**Comment by sidhussmann** [comment]
I downloaded the OEM certificate from [Nuvoton](https://www.nuvoton.com/security/NTC-TPM-EK-Cert/Nuvoton%20TPM%20Root%20CA%202111.cer), converted it to PEM and added it to `/etc/safeboot/certs`. After that I ran a `sudo c_rehash /etc/safeboot/certs/` and did the whole tl;dr procedure from scratch.

I also removed any necessary complexity/potential error sources, such as using `init-key` instead of `yubikey-init` and without `dm-verity`. 

The result is the same: `Unable to verify PCR signature`. 

Eyeballing at https://github.com/osresearch/safeboot/blob/e49799487c37b682cb72e52f4feddc64c3069f62/functions.sh#L188-L196  it is the TPM that fails to verify the PCR signature...

**Comment by sourceXORapprentice** [comment]
Any chance a solution/workaround was reached?

“Unable to verify PCR signature” followed by “TPM TOTP FAILED linux” is frustrating me as well folowing tl;dr installation. I can open a separate issue but the above looks very similar to my experience. 

Installed safeboot-0.7 on Ubuntu 20.04.1 LTS with key-init (no yubikey/SIP/dmeverity), now secureboot is working as intended blocking MSFT keys from a USB boot and it can boot into safeboot recovery. Booting “linux” throws the errors above and requires the recovery key to boot. I reset & updated my firmware and cleaned the TPM after a fresh install with the same results.

Device: HP SFF Elitedesk 705 G3 with an AMD chip. The TPM2.0 module is embedded, reported in dmesg with a generic “MSFT0101:00” while fwupdmgr get-devices provides Infineon (covered I think by /etc/safeboot/certs/IFX_TPM* ?):
├─TPM:
│ │   Device ID:           c6a80ac3a22083423992a3cb15018989f37834d6
│ │   Current version:     7.61.10.57600
│ │   Vendor:              Infineon (TPM:IFX)
│ │   GUIDs:               ff71992e-52f7-5eea-94ef-883e56e034c6 ← system-tpm
│ │                        5eebb112-75ad-5536-b173-a11eb3399402 ← TPM\VEN_IFX&DEV_0000
│ │                        ddf995da-1b32-5a8a-bc1b-8d5af4b38b51 ← TPM\VEN_IFX&MOD_SLB9670
│ │                        6d81ab63-db2e-50ac-934f-6be9accf5e02 ← TPM\VEN_IFX&DEV_0000&VER_2.0
│ │                        301555de-680d-5ddc-b995-7553fc9138f1 ← TPM\VEN_IFX&MOD_SLB9670&VER_2.0
│ │   Device Flags:        • Internal device
│ │ 
│ └─Event Log:
│       Device ID:         58bd405f31c48e6eca290b425f530a94c91e955c
│       Vendor:            Infineon (TPM:IFX)
│       GUID:              a25657fe-b5dc-5be0-8b78-8b9dfec678ff ← system-tpm-eventlog
│       Device Flags:      • Internal device

**Comment by sidhussmann** [comment]
@sourceXORapprentice  so far I haven't gotten time to work further on the T490 problem. The only working device I have, is a little fan-less Lenovo ThinkCentre M90n with an STM TPM2 chip. 

My tl;dr without dm-verity, no sealing pin and no yubikey:

```
# install
git clone https://github.com/osresearch/safeboot.git
cd safeboot
sudo make requirements
make all
make package
cd ..
sudo apt install ./safeboot_0.8_amd64.deb

# init
sudo bash -c 'echo "SEAL_PIN=0" >> /etc/safeboot/local.conf'
sudo safeboot key-init /CN=test/
sudo update-initramfs -u
sudo safeboot uefi-sign-keys
sudo rm -rf /boot/efi/EFI/linux
sudo rm -rf /boot/efi/EFI/recovery
sudo safeboot recovery-sign
sudo safeboot recovery-reboot
sudo reboot

# Should reboot into the recovery image. Login as usual.
sudo safeboot luks-seal
sudo update-initramfs -u
sudo safeboot recovery-sign
sudo safeboot recovery-reboot
sudo reboot

# Should reboot into the recovery image again, with `/` mounted read-only.
sudo safeboot linux-sign
sudo reboot

# Reboot one more time, now into dmverity protect Linux image...
sudo safeboot luks-seal
sudo reboot
```

**Comment by sourceXORapprentice** [comment]
Any reason why this pcrs-relevant line wasn't run - is it in linux-sign now? I had run it on my first attempts and still hit the snags:
sudo safeboot pcrs-sign


Thanks for the tips! I'll keep tinkering. Ran out of time today, and hit secureboot issues after a few hours on a fresh run on Ubuntu Desktop - not server like my first attempts - and hit a lot of new issues. The “sudo rm -rf /boot/efi/EFI/linux” tip saved me from the filespace issue in /boot/efi/. Some additional pre-requisites I had to manually install otherwise it was error city on self-compiled version 0.8:
sudo apt install make git pcsc-tools pcscd xxd openssl tpm2-tools libtss2-dev cryptsetup-bin cryptsetup

My new challenge is I cannot seem to secureboot into the newly signed system like I did before. Obviously a step back from the first several attempts. "sbverify --list linux.efi" shows the /CN=test/ key I had signed the image with, but “sbverify linux.efi” failed verification so despite “sudo safeboot uefi-sign-keys” running it might not be sticking.

There's an option in my HP Setup Menu to manually import Secure Boot keys as der-formatted PK.bin, KEK.bin, DB.bin, DBX.bin in \EFI\HP (http://h10032.www1.hp.com/ctg/Manual/c05649759). Safeboot pumps out .auth files though, not sure how to convert so need to do more research. Secureboot was definitely in setup mode and safeboot uefi-sign-keys ran without complaints. (HP secure boot setup mode: https://youtu.be/5pTBDo95bNs)

**Comment by sourceXORapprentice** [comment]
Got it working now on HP with Infineon TPM2, hoping I can get a flow together in case it can help others.
This overcame “Unable to verify PCR signature” followed by “TPM TOTP FAILED linux” as well as tpm2-tss errors during the luks-seal step by apparently just doing the command twice (?).

Unclear how, but ideas provided here. Possibly the TPM and PCRS entries are really unstable. Potentially may need to ask in [tpm2-tss](https://github.com/tpm2-software/tpm2-tss). 

Two videos provided, of failure and then success. Both were after a fresh Ubuntu Server 20.04.1 install and compiled/installed safeboot 0.8 from source on 1/23/2021 and following tl;dr above:
1. [Fail video](https://youtu.be/KsATiv6lQlI) - Reset TPM in the F10 HP bios/uefi setup menu and set to ask to confirm. It did not. Ran confirm.ran luks-seal once.
2. [Success video](https://youtu.be/h4BBx6ioA8o) - Unseal happens at 9:56, F10 on boot to enter bios setup. Reset TPM with HP bios set to F1 to confirm. F9 to select boot entry for recovery. Ran luks-seal once, appears success but throws many errors. Ran luks-seal again, worked no complaints. Ran pcrs-sign and did linux-sign again in case. It successfully booted linux.efi and unsealed the TPM after reboot and after full shutdown. Continues to work after leaving the machine off several minutes.

Ubuntu Desktop I did not figure out why secure boot kept failing. I started over on Ubuntu Server and secure boot happily took the uefi-sign-keys step and booted linux/recovery with no issues - much like my first time. 

Hoping to do a full run sometime from the start to get a working flow together, apologies the above is muddled or for any inaccuracies, just wanted to document it while it's fresh.

**Comment by anhnd** [comment]
> After attempted boot:
> [tpm-eventlog-after-attempted-boot.log](https://github.com/osresearch/safeboot/files/5659232/tpm-eventlog-after-attempted-boot.log)

@sidhussmann  After reading this, I think your PCR4 wasn't predicted correctly (It involves an EV_EFI_ACTION event, an EV_SEPARATOR event and an EV_EFI_BOOT_SERVICES_APPLICATION event). A quick fix could be achieved by editing this line:

https://github.com/osresearch/safeboot/blob/e49799487c37b682cb72e52f4feddc64c3069f62/sbin/safeboot#L518

to:

```
        ev_act="$(echo -n "Calling EFI Application from Boot Option" | tpm2_trial_extend 0)"
        ex_ev_sep="$(echo "00000000" | hex2bin | tpm2_trial_extend "$ev_act")"
        pcr4_computed="$(echo -n "${ex_ev_sep}${linux_hash}" | hex2bin | sha256)"
```

**Comment by osresearch** [comment]
Ideally for predicting PCRs we would have an easy way to take the eventlog, figure out which combination of actions, separators, and applications are invoked. It's slightly tricky since booting into a recovery kernel via the boot menu changes this path, so we need a non-recovery boot to figure it out (or we need to construct an `OR` policy that would allow us to predict both possibilities).

Nico has just posted a proposal for making it easier to build these complex policy statements #133 that might be a good starting point.

**Comment by anhnd** [comment]
> Ideally for predicting PCRs we would have an easy way to take the eventlog, figure out which combination of actions, separators, and applications are invoked. It's slightly tricky since booting into a recovery kernel via the boot menu changes this path, so we need a non-recovery boot to figure it out (or we need to construct an `OR` policy that would allow us to predict both possibilities).
> 
> 
> 
> Nico has just posted a proposal for making it easier to build these complex policy statements #133 that might be a good starting point.

That would make safeboot a lot more better and I would love to see that. I've been using safeboot for a few days and I've already loved it

**Comment by sidhussmann** [comment]
@anhnd so it seems like `update-initramfs` is struggling with Kernel 5.11. I reverted to Kernel 5.4 and now the DevicePath is pointing correctly to `\EFI\linux\linux.efi`:
[tpm-eventlog-after-attempted-boot.log](https://github.com/osresearch/safeboot/files/7325272/tpm-eventlog-after-attempted-boot.log)
However, after going through all the steps it still fails to verify PCR signature...

Here is my current working branch: https://github.com/sidhussmann/safeboot/tree/47-fix-t490-nuvoton-tpm

**Comment by sidhussmann** [comment]
> oh yes. With kernel 5.11, I did try with Proxmox 7.0 and I had to modified initramfs/hooks/safeboot-hooks (This file is copied to /etc/initramfs-tools/hooks/safeboot-hooks) because some tools might have different names. Basically, I did multiple trial and error attempts of commenting lines and ran "update-initramfs -u" until I found tools with wrong names and fixed them to have a working version. (For example: with Proxmox 7.0, I had to replace copy_exec /usr/bin/xxd with copy_exec /usr/bin/xxd /sbin/)

@anhnd oh cool that you got it working with Proxmox 7.0! That probably means you'd support Debian 11 as well... Have you published your fixes to the initramfs safeboot-hooks? 

> For Kernel 5.4: I had to enter my recovery phase the first time it booted then I ran "sudo safeboot luks-seal", "sudo safeboot linux-sign", "sudo safeboot pcrs-sign" again to make it work. If it won't work after doing that, can you attach a screenshot of PCR values at boot time and your latest tpm event log for comparing with each other? I really believe we can make it work on your laptop!

I just booted again and ran `sudo safeboot luks-seal`, `sudo safeboot linux-sign` and `sudo safeboot pcrs-sign`, as you suggested. Here my screenshot after the following reboot:

![image](https://user-images.githubusercontent.com/37871887/136964987-77027ef0-8b71-4c02-a084-d9f243ade521.png)

**Comment by anhnd** [comment]
> 
> @anhnd oh cool that you got it working with Proxmox 7.0! That probably means you'd support Debian 11 as well... Have you published your fixes to the initramfs safeboot-hooks?
> 

I make a minimal, customized version for my own server here: https://github.com/anhnd/safeboot/tree/disk-encryption-only

> 
> I just booted again and ran `sudo safeboot luks-seal`, `sudo safeboot linux-sign` and `sudo safeboot pcrs-sign`, as you suggested. Here my screenshot after the following reboot:
> 
> ![image](https://user-images.githubusercontent.com/37871887/136964987-77027ef0-8b71-4c02-a084-d9f243ade521.png)

Oh sorry, I was sleepy. This is the same with TPM event log, the thing I want to compare with is the computed PCR4 value when you run "sudo safeboot pcrs-sign" to see if it's predicted correctly, so you can run only "sudo safeboot pcrs-sign" now check it.

**Comment by anhnd** [comment]
> > Oh sorry, I was sleepy. This is the same with TPM event log, the thing I want to compare with is the computed PCR4 value when you run "sudo safeboot pcrs-sign" to see if it's predicted correctly, so you can run only "sudo safeboot pcrs-sign" now check it.
> 
> No worries. I really appreciate your help. Here we go:
> 
> ```
> $ sudo safeboot pcrs-sign
> [sudo] password for gf: 
> Starting PCSCD for yubikey support
> 00000000 [140133358835648] utils.c:81:GetDaemonPid() Can't open /run/pcscd/pcscd.pid: No such file or directory
> WARN: Reading full size of the NV index
> /boot/efi/EFI/linux/linux.efi: TPM version 0000000000000026
> sha256:
>   0 : 0xC05E7C51ED29DED0D50FF55307BC2D7D5FF8B4582BBECD3190D45E69C98E648A
>   2 : 0x3D458CFE55CC03EA1F443F1562BEEC8DF51C75E14A9FCF9A7234A13F198E7969
>   4 : 0x00954CD2CBC55994766E89C267FB32A7AF51A098FB9E092C9946568806CC11E6
>   5 : 0xD613011933404BD03ABB5777935E3E1478B7B7F50932BB610ED225613315C34D
>   7 : 0x6F8F9E7CD8C732F7246EF83AF97A37C745C423FF7251C3F1BCCBE7FA474607C9
> warning: data remaining[111140344 vs 111149960]: gaps between PE/COFF sections?
> /boot/efi/EFI/linux/linux.efi: PE hash 7b780dfc23cd94739bce45325fdd76ffe337562a86e3d4d5e4a31bc9ed9bad61
> PCR4 unchanged
> mode=linux PCR14=4cc49932dc91c7021e5cfee8231a5f2da3ac1de6df7a7aeff333cc8dfd230f28
> final PCRs:
> c05e7c51ed29ded0d50ff55307bc2d7d5ff8b4582bbecd3190d45e69c98e648a
> 3d458cfe55cc03ea1f443f1562beec8df51c75e14a9fcf9a7234a13f198e7969
> 00954cd2cbc55994766e89c267fb32a7af51a098fb9e092c9946568806cc11e6
> d613011933404bd03abb5777935e3e1478b7b7f50932bb610ed225613315c34d
> 6f8f9e7cd8c732f7246ef83af97a37c745c423ff7251c3f1bccbe7fa474607c9
> 4cc49932dc91c7021e5cfee8231a5f2da3ac1de6df7a7aeff333cc8dfd230f28
> Using TPM counter 0000000000000026
> Enter pass phrase for /etc/safeboot/signing.key:02731754 [140133358835648] ccid_usb.c:1264:ControlUSB() control failed (1/2): -7 LIBUSB_ERROR_TIMEOUT
> 
> /sys/firmware/efi/efivars/SafebootPCR-8620893e-c793-457e-8a02-41fc83eef3ce: writing new value
> ```

At least, PCR4(PCR4 unchanged) is predicted correctly now. The only weird part which I don't have is: "02731754 [140133358835648] ccid_usb.c:1264:ControlUSB() control failed (1/2): -7 LIBUSB_ERROR_TIMEOUT" - after looking at your result, I don't think you are using yubikey?

**Comment by sidhussmann** [comment]
> At least, PCR4(PCR4 unchanged) is predicted correctly now. The only weird part which I don't have is: "02731754 [140133358835648] ccid_usb.c:1264:ControlUSB() control failed (1/2): -7 LIBUSB_ERROR_TIMEOUT" - after looking at your result, I don't think you are using yubikey?

No, to limit possible error sources I tried to avoid any unnecessary complexity. I even disabled the sealing pin, which worked on Lenovo ThinkCentre M90n with an STM TPM2 chip.

Here is my current playbook:

```
# first set Secure Boot Setup Mode in BIOS

# install
git clone https://github.com/sidhussmann/safeboot.git
cd safeboot
git checkout 47-fix-t490-nuvoton-tpm
sudo make requirements
make all
make package
sudo dpkg -i safeboot-unstable.deb
sudo apt install -f

# init
sudo bash -c 'echo "SEAL_PIN=0" >> /etc/safeboot/local.conf'
sudo safeboot key-init /CN=test/
sudo update-initramfs -u
sudo safeboot uefi-sign-keys
sudo safeboot recovery-sign
sudo safeboot recovery-reboot
sudo reboot

# Should reboot into the recovery image. Login as usual.
sudo safeboot luks-seal
sudo safeboot recovery-sign
sudo safeboot recovery-reboot
sudo reboot

# Should reboot into the recovery image again, with `/` mounted read-only.
sudo safeboot linux-sign
sudo reboot

# Reboot one more time, now into dmverity protect Linux image...
sudo safeboot luks-seal
sudo reboot
```
